requestMtu return is missing await keyword

efreibe opened the pull request:
In some cases write operations will fail, sending data will be started before setting mtu value, because the promise is never awaited.

efreibe committed:
requestMtu return is missing await keyword

In some cases write operations will fail, sending data will be started before setting mtu value, because the promise is never awaited.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pauldemarco/flutter_blue?pullRequest=579) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pauldemarco/flutter_blue?pullRequest=579) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pauldemarco/flutter_blue?pullRequest=579) it.</sub>

efreibe commented:
That PR was not working after all for me, I needed to add some code to make it work properly, please review code below, I think it will need a final touch to work properly.

```dart
Future<void> requestMtu(int desiredMtu) async {
    var request = protos.MtuSizeRequest.create()
      ..remoteId = id.toString()
      ..mtu = desiredMtu;

    await FlutterBlue.instance._channel.invokeMethod('requestMtu', 
      request.writeToBuffer());

    await FlutterBlue.instance._methodStream
      .where((m) => m.method == "MtuSize")
      .map((m) => m.arguments)
      .map((buffer) => protos.MtuSizeResponse.fromBuffer(buffer))
      .where((p) => p.remoteId == id.toString())
      .map((p) => p.mtu)
      .first;
  }
```

efreibe committed:
Adjustment needed

Mtu changes send back a response, this response needs to be processed after leaving the method in flutter.

slecoustre commented:
Hello
Can you tell me when that Pull Request is going to be merged?

thanks

madsdyd commented:
Got this issue today as well (I think): Setting MTU prior to e.g. characteristics discovery seems to ignore the await, and mess up with e.g. characteristics discovery.

Had to move the MTU change to somewhere where no other BLE activity was ongoing. Then worked as expected.

slecoustre commented:
Hello, this means that the code won't be merged or has been fixed on another branch.

savar commented on lib/src/bluetooth_device.dart:
```diff
@@ -128,8 +128,16 @@ class BluetoothDevice {
       ..mtu = desiredMtu;
 
-    return FlutterBlue.instance._channel
+    await FlutterBlue.instance._channel
```
as I am new to dart I am not sure if it is true what I write, but is this really making any difference?

you do at the moment within the async method:

```dart
await FlutterBlue.instance._channel.invokeMethod('...');
await FlutterBlue.instance._methodStream.where('.. MtuSize...')
```
while as the old version was returning a future you could have done from the outside
```dart
await btDevice.requestMtu(111); // which is like your first await within the `requestMtu()` method
await btDevice.mtu.first; // which is like your second await within the `requestMtu()` method
```

so I might be wrong (seriously!) but I don't see that this change actually changes anything

efreibe replied:
Yes, you are right, but that code is incomplete, I've sent a complete PR right after the first, and you absolutely need the await keyword in that code.

https://github.com/pauldemarco/flutter_blue/pull/579#issuecomment-628325875

efreibe replied:
Now I see that i've missed a return in last code, you can add it, but it is working for me as is.

savar replied:
why don't you update your pull request then? I was (and still am) a little bit confused about a pull request which has updates in comments :thinking:

savar replied:
besides that, I still don't get it then (which might be, as I am still confused about darts async style). I thought the
```dart
Future<void> requestMtu(int desiredMtu) async {
   // var stuff..

   return FlutterBlue.instance._channel.invokeMethod('requestMtu', request.writeToBuffer());
}
```
and then doing in another
```dart
Future<void> someMethod() async {
  await this.btDevice.requestMtu(123);
}
```
is the same as doing it simply in your method.. and the second part you can do as well in your method.. I still don't see it or the async stuff is not working (for me) as I would it expect to be.

efreibe replied:
I don't know how pull request works in github, I've never used that, I've posted my code (and repairs I've done) to help other users to benefit from that, but I couldn't change my original (and useless) PR so I left a response below that. 
I thought that what I've done was enough for the author review and integrate that code into the repo. 

As for the second question, there is nothing weird about ```await``` or ```async```, the fact is that the function does not end in:

```dart
return FlutterBlue.instance._channel.invokeMethod('requestMtu', request.writeToBuffer());
``` 

Invoking ```requestMtu``` will call this function inside android, then it will wait a confirmation event. So you must request the MTU and wait for the stream channel to return when the MTU has changed, if you don't add this code it won't work:

```dart
    await FlutterBlue.instance._channel.invokeMethod('requestMtu', 
      request.writeToBuffer());

    await FlutterBlue.instance._methodStream
      .where((m) => m.method == "MtuSize")
      .map((m) => m.arguments)
      .map((buffer) => protos.MtuSizeResponse.fromBuffer(buffer))
      .where((p) => p.remoteId == id.toString())
      .map((p) => p.mtu)
      .first;
```

I've copied and adapted that code from another place inside the sources, so I asked to validate the code before commiting any changes.

Finally, you can use promises of course:
```dart
    return FlutterBlue.instance._channel.invokeMethod('requestMtu', request.writeToBuffer()).then(
      (_) => FlutterBlue
        .instance._methodStream
        .where((m) => m.method == "MtuSize")
        .map((m) => m.arguments)
        .map((buffer) => protos.MtuSizeResponse.fromBuffer(buffer))
        .where((p) => p.remoteId == id.toString())
        .map((p) => p.mtu)
        .first);
```

and await outside:
```dart
Future<void> someMethod() async {
  await this.btDevice.requestMtu(123);
}
```

savar replied:
I think I get the issue now and it is mainly not really in this code per se.. your solution is one way. The other would be to fix
```dart
  /// The MTU size in bytes
  Stream<int> get mtu async* {
    yield await FlutterBlue.instance._channel
        .invokeMethod('mtu', id.toString())
        .then((buffer) => new protos.MtuSizeResponse.fromBuffer(buffer))
        .then((p) => p.mtu);

    yield* FlutterBlue.instance._methodStream
        .where((m) => m.method == "MtuSize")
        .map((m) => m.arguments)
        .map((buffer) => new protos.MtuSizeResponse.fromBuffer(buffer))
        .where((p) => p.remoteId == id.toString())
        .map((p) => p.mtu);
  }
```

as this one returns the "right now MTU" initially.. that's why you cannot wait properly for the change.. removing the first `yield await` would make it possible to simply do

```dart
await device.requestMtu(247);
await device.mtu.first;
```

Bobby9292 commented:
Any idea if this modification will be release ?
It is a pretty critical issue on the lib ATM.

savar committed:
using the same style `discoverServices()`

* to ensure that we don't miss the call from the response we should have
  the listener setup before calling the `requestMtu` internal function
* instead of using the stream just for locking, we can also return the
  value
* awaiting the final call doesn't make any difference so its clearer to
  return it as the caller needs to `await` anyways

savar commented:
@efreibe please check https://github.com/efreibe/flutter_blue/pull/1 and tell me what you think.. in case you merge it I would expect (but to be honest I am unsure and never tried that) that this pull request will be updated as well!

savar commented:
actually after writing about that, there is one way to overcome the issue for the time being without patching anything:

```dart
// skipping the 1st as this is the "actual" mtu and not the changed one
device.mtu.elementAt(1).then((mtu) {
   mtu = mtu < 23 ? 20 : mtu - 3; // failsafe by always assuming a ATT MTU and not a DATA MTU
   // do your service discovery
  this.serviceDiscovery();
}
await device.requestMtu(100);
```

with this, no patch required even though I think it makes sense to patch it

efreibe committed:
Merge pull request #1 from savar/patch_the_patch

using the same style `discoverServices()`

a10waveracer commented:
Just wanted to chime in and say that this PR was super helpful in fixing an issue. Thank you!

toxichand commented:
It really helps me to overcome this issue, thank you!

flukejones commented:
This issue and PR is coming up to a one year anniversary..

Because of the lack of ongoing maintenance and piling up of issues I think I'll be moving back to the Polidea/Phillips lib

akakiev commented:
> actually after writing about that, there is one way to overcome the issue for the time being without patching anything:
> 
> ```dart
> // skipping the 1st as this is the "actual" mtu and not the changed one
> device.mtu.elementAt(1).then((mtu) {
>    mtu = mtu < 23 ? 20 : mtu - 3; // failsafe by always assuming an ATT MTU and not a DATA MTU
>    // do your service discovery
>   this.serviceDiscovery();
> }
> await device.requestMtu(100);
> ```
> 
> with this, no patch required even though I think it makes sense to patch it

Hello, can you explain for me, please, what is going on at this code snippet ?

savar commented:
> 
> Hello, can you explain for me, please, what is going on at this code snippet ?

Hey, I am not sure what I can tell you more than what the comments are already doing.
The method `requestMtu()` is not waiting for the actually underlying command to finish, so you don't know when the MTU change actually succeeded/happened. The trick on this code block is, that you wait for the notification of the change. If you use `.mtu` on a device you get back a stream of the mtu. The stream will initially give back the actual MTU which it knows and then you can listen to new MTU changes and react on them.

So the goal was to change the MTU and only after this was _done_, that we wanted to start the discovery (because if you start too early everything goes wrong).

The trick was: listen on the `.mtu` stream, ignore the first one as it is the "actual" (not yet changed) MTU, but wait for the second one. The second one would be the "notification" of the successful `.requestMtu()` command when this was done.

So `device.mtu.elementAt(1).then((mtu) {` will wait for the change and then within the `.then()` part you can do whatever you actually want to do like service discovery or whatever you need.

akakiev commented:
@savar thank you very much, now it is more clear, but one thing still is wondering: why this expression should be done after ignoring the first one MTU - mtu = mtu < 23 ? 20  : mtu - 3; ?

savar commented:
> @savar thank you very much, now it is more clear, but one thing still is wondering: why this expression should be done after ignoring the first one MTU - mtu = mtu < 23 ? 20 : mtu - 3; ?

This `mtu` variable was used internally to know how big a chunk of data can be. Check the BT LE protocol like here: https://punchthrough.com/maximizing-ble-throughput-part-2-use-larger-att-mtu-2/ to see that you normally would have an ATT MTU but as far as I can remember, the MTU was not always actually (depending on the underlying system) including the 3 bytes for the `op-code` and `attribute handle` but was sometimes just `20`. So as a failsafe if we get anything lower than 23 byte, we just say the "DATA MTU" is 20 bytes and we are safe and do not risk anything.

If this is really required, I am not sure anymore, but I do not work on Dart/Flutter anymore nor do we use flutter_blue as its development halted and it misses some features we needed back then (but don't ask me which ones, I cannot remember).

akakiev commented:
@savar thank you. I try to use: await Future.delayed(Duration(milliseconds: 2000)); instead of this code snippet and it works also but with time delay 2 sec, with this code snippet there is no time delaytion, so how do you think it is better - use this code snippet or await Future.delayed() function ?

savar commented:
> @savar thank you. I try to use: await Future.delayed(Duration(milliseconds: 2000)); instead of this code snippet and it works also but with time delay 2 sec, with this code snippet there is no time delaytion, so how do you think it is better - use this code snippet or await Future.delayed() function ?

Hey @akakiev , my code snippet is just more or less doing what you want: request for a change in MTU size and as soon as it is done, you can start doing (probably service discovery) what you want immediately.

With your "Delay" approach you have 2 issues:
1. you always have a delay as it normally doesn't take 2s (bad user experience)
2. in case it will, for whatever reason (bad connection, remote device is behaving erratic, ...), take longer than 2s, then your code will fail and you will have exceptions (probably bad or completely unhandled).

So for me there is no real question which of these 2 variants you should use. But at the end, that's only up to you.

akakiev commented:
thank you, @savar! I have the last question, maybe it will sounds very simple for you, but I am currently only starts my experience with flutter and it's 'futures, async, await, then' : why the row 'await device.requestMtu(128);' at the end of this code snippet, why it is outside of the brackets, but it works before serviceDiscovery() function ? I'm used to sequential order...

savar commented:
> thank you, @savar! I have the last question, maybe it will sounds very simple for you, but I am currently only starts my experience with flutter and it's 'futures, async, await, then' : why the row 'await device.requestMtu(128);' at the end of this code snippet, why it is outside of the brackets, but it works before serviceDiscovery() function ? I'm used to sequential order...

Hey @akakiev , I am sorry but this is going out of hand. This is a merge request and not a learning platform for flutter. You should check out https://dart.dev/codelabs/async-await or something like that, but please stay focused and on the topic in this merge request.

akakiev commented:
thank you, @savar! In continuation of this topic, I created a question on stackoverflow: 
https://stackoverflow.com/questions/76128783/how-to-return-data-from-the-flutter-provider-side-to-the-vue-js-side-via-inappwe
Please, check it, maybe you can answer it.
